Update runtime; tweaks

rossberg opened the pull request:
Various tweaks, updates, and todos to get the ball rolling again. In particular, split store into shared and local parts and switch to symbolic addresses.

conrad-watt commented on document/core/exec/runtime.rst:
```diff
@@ -277,24 +315,26 @@ It is an invariant of the semantics that the length of the element vector never
 Memory Instances
 ~~~~~~~~~~~~~~~~
 
+.. todo:: We used to update the memory type when a memory is grown. This does not work for shared memories. In fact, the "current" type of a shared memory is nondeterministic. We need to model that during instantiation somehow.
```
I had assumed that instantiation would proceed as though executing `memory.size` on each imported memory (similar to the way global initialisers are evaluated), but I realise now that this hasn't really been discussed.

IIUC, the memory type is only introspected during instantiation, so for instruction validation and the soundness statement, it would be OK for the memory type to no longer contain a `min` size?

rossberg replied:
Memory types are used for a number of purposes, and its notion of limits is furthermore shared with tables. I'd rather not introduce multiple different variations and intersections of these concepts. Rather, some form of abstraction to read the current "dynamic" type of a memory at instantiation time. That would also be what's used by the embedder API.

conrad-watt replied:
to be bikeshedded when instantiation is specified

conrad-watt commented on document/core/exec/runtime.rst:
```diff
@@ -277,24 +315,26 @@ It is an invariant of the semantics that the length of the element vector never
    \production{(memory instance)} & \meminst &::=&
-     \{ \MIMAX~\u32^?, \MISHARE~\UNSHARED, \MIDATA~\vec(\byte) \} \\ &&|&
-     \{ \MIMAX~\u32^?, \MISHARE~\SHARED \} \\
+     \{ \MITYPE~\memtype, \MIDATA~\vec(\byte) \} \\ &&|&
```
syntactically there's nothing here stopping memtype from being `shared`, even if there's a byte vector, right?

rossberg replied:
Yes, it's just an invariant of the semantics (I actually had a statement about that initially, maybe I should reinstate it for clarity).

conrad-watt commented on document/core/exec/runtime.rst:
```diff
@@ -702,38 +696,115 @@ Its form and meaning is outside the scope of this specification.
+
+.. math::
+   \begin{array}{@{}c@{}}
+   \XX{SS}; P_1^\ast~(S; \instr^\ast \AT h)~P_2^\ast
```
I might be misreading this rule (I'm having trouble getting the markup building to work) but why is the empty frame `F_\epsilon` needed? I would have thought that each thread would have its own frame (as defined [here](https://webassembly.github.io/spec/core/exec/runtime.html?highlight=thread#configurations)) instead of just being local store + instructions.

rossberg replied:
Well, then you would just permanently keep all the empty frames in the global configuration (and formulate well-formedness rules for them), and you'd need to create them during thread creation.

The main reason I did it like that in the original formulation was that there was only one level of reduction relation, so no natural place to inject the empty frame as you go. But now that we have that, it seems less attractive to keep that artefact around. The only appeal would be that thread and local configuration get unified, but is that valuable? I don't know, though I could be convinced otherwise.

Btw, I don't know if you noticed, but despite the split store, the rule here is currently simpler than the one in our paper:

* It does not allow any allocation, i.e., the global store remains unchanged. At least for now, the only operation that allocates is instantiate, which is simpler to model with a global reduction rule.

* It also doesn't pass the prior global store to the local reduction, since the operations where it would be needed (mutation of a shared table or global) are not currently supported.

So this completely decouples local reduction from the global store for now, which is nice. Even if we add allocation forms, it might actually be simpler to do so through directly through global reductions. (Typed updates will remain an annoyance, though.)

conrad-watt replied:
My thinking was that unifying threads and local configurations would mean they wouldn't need separate validity rules in the soundness statement. It's a very small thing either way.

Rest of your comment looks good!

conrad-watt commented on document/core/appendix/properties.rst:
```diff
@@ -745,10 +745,23 @@ In other words, every thread in a valid configuration either runs forever, traps
 
+.. index:: relaxed memory, ! sequential consistency
+
+Sequential Consistency of Data-Race-Free Programs
```
If you remember the corrigendum I sent you (still going through the ACM process), we have to be a little apologetic about SC-DRF - the property we inherit from JS is weaker than the ideal one (which is technically false, because of a corner to do with thin-air).

I should handle writing something about this - basically a non-normative note _somewhere_ (appendix?) to the effect that because of known limitations in our ability to formalise relaxed memory, the model has gaps in it which are subjects of active research.

conrad-watt replied:
agree

conrad-watt reviewed (commented):
Had an initial look, still trying to get the document to build locally so I can properly see the markup.

The only actionable question I have is about the global reduction rule (see below). Other comments can probably be addressed later in drafting.

rossberg reviewed (commented):
I can't build myself right now, our Sphinx hacks bit-rot too quickly. :( The problems should be fixed with the merge in #167, though.

conrad-watt reviewed (approved):
LGTM then if the document can't be built right now

WDYT about my point regarding the empty frame in the global reduction rule?

rossberg commented:
> WDYT about my point regarding the empty frame in the global reduction rule?

We can always change that later, it doesn't affect much else, I think. Right now my slight preference is to keep it this way, because conceptually, threads do not really have a frame -- that's an attribute of function activations, and a local configuration nicely corresponds to the current activation. Inserting the empty frame in this rule encapsulates this auxiliary device to bridge the gap within a single rule.

I'll wait with landing until #167 is merged and I have rebased this PR.

rossberg commented:
@conrad-watt, I merged upstream and fixed the latex, should build again. Also adjusted rules to the changed meminst def that I overlooked before. PTAL at the last commit.

conrad-watt commented on document/core/exec/modules.rst:
```diff
@@ -325,7 +327,7 @@ New instances of :ref:`functions <syntax-funcinst>`, :ref:`tables <syntax-tablei
 3. Let :math:`a` be the first free :ref:`memory address <syntax-memaddr>` in :math:`S`.
 
-4. Let :math:`\meminst` be the :ref:`memory instance <syntax-meminst>` :math:`\{ \MIDATA~(\hex{00})^{n \cdot 64\,\F{Ki}}, \MIMAX~m^?, \MISHARE~s\}` that contains :math:`n` pages of zeroed :ref:`bytes <syntax-byte>`.
+4. Let :math:`\meminst` be the :ref:`memory instance <syntax-meminst>` :math:`\{ \MITYPE~\memtype, \MIDATA~(\hex{00})^{n \cdot 64\,\F{Ki}} \}` that contains :math:`n` pages of zeroed :ref:`bytes <syntax-byte>`.
```
Do you think it's worth changing this now so that the `\MIDATA` component is only present if the memory type is `unshared`, or is that part of the above todo?

rossberg replied:
Yes, I think adapting this requires a larger refactoring. For now I just adjusted the syntax to be consistent (and to avoid Latex errors due to missing defs).

conrad-watt commented on document/core/exec/modules.rst:
```diff
@@ -337,7 +339,7 @@ New instances of :ref:`functions <syntax-funcinst>`, :ref:`tables <syntax-tablei
    \memtype &=& \{\LMIN~n, \LMAX~m^?\}~s \\
    \memaddr &=& |S.\SMEMS| \\
-   \meminst &=& \{ \MIDATA~(\hex{00})^{n \cdot 64\,\F{Ki}}, \MIMAX~m^?, \MISHARE~s\} \\
+   \meminst &=& \{ \MITYPE~\memtype, \MIDATA~(\hex{00})^{n \cdot 64\,\F{Ki}} \} \\
```
same comment as above

conrad-watt commented on document/core/exec/relaxed.rst:
```diff
@@ -1,8 +1,30 @@
+Validity
+~~~~~~~~
+
+.. todo:: rename to avoid confusion with validation
```
We could rename "validity" to "consistency". The former is ARM model terminology, the later is C++ model terminology.

rossberg replied:
Consistency sounds good to me! I went ahead and renamed it.

conrad-watt commented on document/core/exec/modules.rst:
```diff
@@ -59,15 +59,17 @@ The following auxiliary typing rules specify this typing relation relative to a
 .. math::
    \frac{
-     S.\SMEMS[a] = \{ \MIDATA~b^{n\cdot64\,\F{Ki}}, \MIMAX~m^?, \MISHARE~s\}
+     S.\SMEMS[a] = \{ \MITYPE~\{\LMIN~n, \LMAX~m^?\}~s, \MIDATA~b^{k\cdot64\,\F{Ki}} \}
```
What's the reasoning for this change? It's a little strange that the memory instance now has a "minimum" component which is never used, and it makes the definition of external typing more surprising. The change is also not currently mirrored in the table instance, although it could be.

It does make the soundness proof somewhat neater (give the related change to memory instance validity). Is that why?

conrad-watt replied:
I just realised the change to the syntax of the memory instance was made before and I missed it - I should have asked this question them

conrad-watt commented on document/core/util/macros.def:
```diff
@@ -1068,17 +1074,19 @@
 
-.. |valid| mathdef:: \mathrel{\mbox{valid}}
-.. |validwith| mathdef:: \mathrel{\mbox{valid-with}}
+.. todo:: include xrefs
```
Is this related to the CI builds failing?

rossberg replied:
Probably. I moved the todo elsewhere.

conrad-watt reviewed (approved):
LGTM!
